### PR TITLE
Added distinguished, date-time-based filenames to reports

### PR DESCRIPTION
**Pull request by markreay**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #1443097
- [x] Added relevant unit test for your changes. (`npm run test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] Ran precheckin (`npm run precheckin`)
- [x] Added screenshots/GIFs for UI changes.

#### Description of changes

No UI changes strictly. But the file names have been changed for both reports as illustrated in the following screenshot.

image

The example file names are 

* AssessmentReport_20190320_AccessibleUniversity.html
* AutomatedChecksReport_20190320_AccessibleUniversity.html

The file names include the date and an abbreviated version of the site title.

#### Notes for reviewers

I leveraged and enhanced existing report name generation code.

### CODE REVIEW

File: src/DetailsView/reports/components/report-file-name-generator.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+type DateProvider = () => Date;
```
Comment (smoralesd) [inline src/DetailsView/reports/components/report-file-name-generator.ts]: we already have something similar to this (a class though, not a type)

File: src/tests/unit/tests/DetailsView/components/export-dialog.test.tsx [hunk 2]
```diff
@@ -4,8 +4,13 @@ import { IMock, Mock, MockBehavior, Times } from 'typemoq';
 
 import { DetailsViewActionMessageCreator } from '../../../../../DetailsView/actions/details-view-action-message-creator';
 import { ExportDialog, ExportDialogProps } from '../../../../../DetailsView/components/export-dialog';
+import { ReportFileNameGenerator } from '../../../../../DetailsView/reports/components/report-file-name-generator';
 
 describe('ExportDialog', () => {
+    const reportFileName = 'THE REPORT FILE NAME';
+    const reportFileNameGeneratorMock = Mock.ofType(ReportFileNameGenerator);
```
Comment (smoralesd) [inline src/tests/unit/tests/DetailsView/components/export-dialog.test.tsx]: you probably want to use `Mock.ofType<ReportFileNameGenerator>();` instead... works fine with interfaces and types, don't need a concrete class

File: src/DetailsView/reports/components/report-file-name-generator.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+export type ReportFileNameGeneratorDeps = {
+    dateProvider: DateProvider;
+};
+
+export class ReportFileNameGenerator {
```
Comment (dbjorge) [inline src/DetailsView/reports/components/report-file-name-generator.ts]: >ReportFileNameGenerator [](start = 13, length = 23)

Discussed offline with mark that we already have a ReportNameGenerator with very similar functionality (but which seems to be dead code right now, that shows up as issues table state that never gets read). Mark's going to look into consolidating

File: src/tests/unit/tests/DetailsView/reports/components/report-file-name-generator.test.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { ReportFileNameGenerator } from '../../../../../../DetailsView/reports/components/report-file-name-generator';
```
Comment (smoralesd) [inline src/tests/unit/tests/DetailsView/reports/components/report-file-name-generator.test.ts]: copyright header should be on top

File: src/DetailsView/components/issues-table.tsx [hunk 5]
```diff
@@ -125,6 +126,7 @@ export class IssuesTable extends React.Component<IssuesTableProps, IssuesTableSt
             <ExportDialog
                 deps={this.props.deps}
                 isOpen={this.state.isExportDialogOpen}
+                fileNameBase="AutomatedChecksReport"
```
Comment (peterdur) [inline src/DetailsView/components/issues-table.tsx]: It looks like when I converged both reports on Export Dialog as part of the OF6 update, I [didn't wire up](https://github.com/Microsoft/accessibility-insights-web/commit/a1bc41e5b6247a30b16d634af8e04ec3cf6a3fc3#diff-0d3bccd06335566e1695cedac734218cL158
) the generated file name for the automated checks report. That's actually still available here at `this.state.exportName`. It contains the date (not the time) from the report and the target page name. There might be something reusable here.

File: src/DetailsView/reports/components/report-file-name-generator.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    constructor(private readonly deps: ReportFileNameGeneratorDeps) {}
+
+    public getFileName(baseName: string, extension: string): string {
+        const { dateProvider } = this.deps;
+        const isoString = dateProvider().toISOString();
+        const dateTime = `${isoString.substr(0, 10)} ${isoString.substr(11, 8)}`;
```
Comment (peterdur) [inline src/DetailsView/reports/components/report-file-name-generator.ts]: I noticed that this generates a GMT date/time for the file name, whereas the date/time shown in the report is local time. Should the two match?

Comment (peterdur) [reply src/DetailsView/reports/components/report-file-name-generator.ts]: Another reason the timestamps might diverge for automated checks between the report and the file name: the report shows the time of the scan, and the file name is the time of the export. (Assessment reports currently display the time of report generation, so they'd only be different if a displayed time boundary is crossed.)

File: src/DetailsView/components/issues-table.tsx (hunk 5 above)
Comment (markreay) [reply src/DetailsView/components/issues-table.tsx]: I believe I have wired it up as you expected. Let me know if it isn't right.

---
In reply to: [267592312](https://github.com/Microsoft/accessibility-insights-web/pull/398#discussion_r267592312) [](ancestors = 267592312)

File: src/tests/unit/tests/DetailsView/components/export-dialog.test.tsx (hunk 2 above)
Comment (markreay) [reply src/tests/unit/tests/DetailsView/components/export-dialog.test.tsx]: I removed that mock so the comment isn't relevant anymore. I noted your feedback though, thanks.

---
In reply to: [267545106](https://github.com/Microsoft/accessibility-insights-web/pull/398#discussion_r267545106) [](ancestors = 267545106)

File: src/DetailsView/reports/components/report-file-name-generator.ts (hunk 3 above)
Comment (markreay) [reply src/DetailsView/reports/components/report-file-name-generator.ts]: I consolidated.

File: src/tests/unit/tests/DetailsView/reports/components/report-file-name-generator.test.ts (hunk 4 above)
Comment (markreay) [reply src/tests/unit/tests/DetailsView/reports/components/report-file-name-generator.test.ts]: File removed.

File: src/DetailsView/reports/components/report-file-name-generator.ts (hunk 6 above)
Comment (markreay) [reply src/DetailsView/reports/components/report-file-name-generator.ts]: Back to local time.

File: src/DetailsView/components/export-dialog.tsx [hunk 7]
```diff
@@ -8,10 +8,12 @@ import * as React from 'react';
 
 import { ExportResultType } from '../../common/telemetry-events';
 import { DetailsViewActionMessageCreator } from '../actions/details-view-action-message-creator';
+import { ReportNameGenerator } from '../reports/report-name-generator';
```
Comment (peterdur) [inline src/DetailsView/components/export-dialog.tsx]: This import is no longer needed.

File: src/DetailsView/reports/report-generator.ts [hunk 8]
```diff
@@ -18,24 +18,22 @@ export class ReportGenerator {
...
 
     public generateAssessmentHtml(
-        deps: ReportGeneratorDeps,
         assessmentStoreData: IAssessmentStoreData,
         assessmentsProvider: IAssessmentsProvider,
         featureFlagStoreData: FeatureFlagStoreData,
         tabStoreData: ITabStoreData,
         description: string,
     ): string {
         return this.assessmentReportHtmlGenerator.generateHtml(
```
Comment (peterdur) [inline src/DetailsView/reports/report-generator.ts]: In a more consistent world, we'd pass the `scanDate` to `assessmentReportHtmlGenerator.generateHtml` to ensure that the same date is used in both places (and from there to `assessmentReportModelBuilderFactory.create`) - but given that we're now only using the date (in local time zone) in the file name, there's not much of a chance of a user-visible mismatch. Given that this is an existing inconsistency between the assessment and the automated checks report code, not a blocker.

File: src/DetailsView/components/export-dialog.tsx (hunk 7 above)
Comment (markreay) [reply src/DetailsView/components/export-dialog.tsx]: Removed.

File: src/DetailsView/reports/report-generator.ts (hunk 8 above)
Comment (markreay) [reply src/DetailsView/reports/report-generator.ts]: Agreed. Discussed offline with @peterdur and we will save this for the future.

**Pull request merged**